src/wallet/wallet.cpp

```diff
@@ -964,15 +964,30 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFromLoadWallet, CWalletD
 
 bool CWallet::UpdatedNoteData(const CWalletTx& wtxIn, CWalletTx& wtx)
 {
+    // On reindex, IncrementNoteWitnesses() is called for every block again. So
+    // we drop the cached witnesses for any transactions we already know about.
+    bool dropCache = GetBoolArg("-reindex", false);
```

Comment:
Also this is looking at global state so it is not as easily unit-testable. Would it make more sense to pass in ``dropCache`` from the caller?